hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.state;
+
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.SqlResultImpl;
+import com.hazelcast.sql.impl.client.SqlPage;
+import com.hazelcast.sql.impl.client.SqlPageRow;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Registry of active client cursors.
+ */
+public class QueryClientStateRegistry {
+
+    private final ConcurrentHashMap<QueryId, QueryClientState> clientCursors = new ConcurrentHashMap<>();
+
+    public SqlPage registerAndFetch(
+        UUID clientId,
+        SqlResultImpl cursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        QueryClientState clientCursor = new QueryClientState(clientId, cursor);
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (!page.isLast()) {
+            clientCursors.put(cursor.getQueryId(), clientCursor);
+        }
+
+        return page;
+    }
+
+    public SqlPage fetch(
+        UUID clientId,
+        QueryId queryId,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        QueryClientState clientCursor = getClientCursor(clientId, queryId);
+
+        if (clientCursor == null) {
+            throw QueryException.error("Query cursor is not found (closed?): " + queryId);
+        }
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (page.isLast()) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return page;
+    }
+
+    private SqlPage fetchInternal(
+        QueryClientState clientCursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        Iterator<SqlRow> iterator = clientCursor.getIterator();
+
+        List<SqlPageRow> page = new ArrayList<>(cursorBufferSize);
+        boolean last = fetchPage(iterator, page, cursorBufferSize, serializationService);
+
+        if (last) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return new SqlPage(page, last);
+    }
+
+    private static boolean fetchPage(
+        Iterator<SqlRow> iterator,
+        List<SqlPageRow> page,
```

Comment:
Would it be possible to use List<Data> instead of SqlPageRow here. And also in the protocol ?  
It would eliminate the need for creating a new type in the protocol, as well as creating extra objects here. 
In the protocol, we may also eliminate SqlPage in the protocol and put the `last boolean` as one of the response params.

Reply:
Removed `SqlPage` from the protocol, although it is still used in the server-side part, because we need to return rows and a flag anyway.